AsyncGuidance.md

```diff
@@ -839,6 +840,144 @@ class AmbientValues
 
 :white_check_mark: **GOOD** The above uses a `ConcurrentDictionary<int, string>` which is thread safe.
 
+### Don't leak your AsyncLocal\<T\>
+
+Async locals flow across awaits automatically and can be captured by any API that explicitly calls `ExecutionContext.Capture`. The latter can lead to memory leaks if the lifetime of the API capturing the execution context is incompatible with the lifetime of the data stored in the async local.
+
+#### Common APIs that capture the ExecutionContext
+
+APIs that run user callbacks usually store capture the current execution context in order to preserve async locals between callback registration and execution. Here are examples of some APIs that do this:
+
+- `Timer`
+- `CancellationToken.Register`
+- `new FileSystemWatcher`
+- `SocketAsyncEventArgs`
+- `Task.Run`
+- `ThreadPool.QueueUserWorkItem`
+
+❌ Here's an example of an execution context leak that causes memory pressure.
```

Comment:
```suggestion
❌ Here's an example of an execution context leak that causes memory pressure because of a lifetime mismatch between the API capturing the execution context and the lifetime of the data stored in the async local.
```